src/Mvc/test/Mvc.IntegrationTests/FormFileModelBindingIntegrationTest.cs

```diff
@@ -354,15 +747,491 @@ public async Task BindProperty_WithData_WithPrefix_GetsBound()
+            var kvp = Assert.Single(modelState);
+
+            Assert.Equal("house.Garage.Cars[0].Specs", kvp.Key);
+        }
+
+        [Fact]
+        public async Task BindProperty_FormFile_AsAPropertyOnNestedColection()
+        {
+            // Arrange
+            var parameterBinder = ModelBindingTestHelper.GetParameterBinder();
+            var parameter = new ParameterDescriptor
+            {
+                Name = "p",
+                BindingInfo = new BindingInfo(),
+                ParameterType = typeof(Car1)
+            };
+
+            var data = "Some Data Is Better Than No Data.";
+            var testContext = ModelBindingTestHelper.GetTestContext(
+                request =>
+                {
+                    request.QueryString = QueryString.Create("p.Name", "Accord");
+                    UpdateRequest(request, data, "p.Specs");
+                });
+
+            var modelState = testContext.ModelState;
+
+            // Act
+            var modelBindingResult = await parameterBinder.BindModelAsync(parameter, testContext);
+
+            // Assert
+
+            // ModelBindingResult
+            Assert.True(modelBindingResult.IsModelSet);
+
+            // Model
+            var car = Assert.IsType<Car1>(modelBindingResult.Model);
+            Assert.NotNull(car.Specs);
+            var file = Assert.Single(car.Specs);
+            Assert.Equal("form-data; name=p.Specs; filename=text.txt", file.ContentDisposition);
+            var reader = new StreamReader(file.OpenReadStream());
+            Assert.Equal(data, reader.ReadToEnd());
+
+            // ModelState
+            Assert.True(modelState.IsValid);
+            Assert.Equal(2, modelState.Count);
+
+            var entry = Assert.Single(modelState, e => e.Key == "p.Name").Value;
+            Assert.Equal("Accord", entry.AttemptedValue);
+            Assert.Equal("Accord", entry.RawValue);
+
+            Assert.Single(modelState, e => e.Key == "p.Specs");
+        }
+
+        public class MultiDimensionalFormFileContainer
+        {
+            public IFormFile[][] FormFiles { get; set; }
+        }
+
+        [Fact]
+        public async Task BindModelAsync_MultiDimensionalFormFile_Works()
+        {
+            // Arrange
+            var parameterBinder = ModelBindingTestHelper.GetParameterBinder();
+            var parameter = new ParameterDescriptor
+            {
+                Name = "p",
+                BindingInfo = new BindingInfo(),
+                ParameterType = typeof(MultiDimensionalFormFileContainer)
+            };
+
+            var data = "Some Data Is Better Than No Data.";
+            var testContext = ModelBindingTestHelper.GetTestContext(
+                request =>
+                {
+                    UpdateRequest(request, data + 1, "FormFiles[0]");
+                    AddFormFile(request, data + 2, "FormFiles[1]");
+                    AddFormFile(request, data + 3, "FormFiles[1]");
```

Comment:
Does anything different happen if the files are explicitly named e.g. "FormFiles[0][0]", "FormFiles[1][0]" and "FormFiles[1][1]"?